lib/logstash/plugin_mixins/scheduler.rb

```diff
@@ -0,0 +1,60 @@
+# encoding: utf-8
+
+require 'logstash/namespace'
+require 'logstash/plugin'
+
+require 'logstash/plugin_mixins/scheduler/rufus_impl'
+
+module LogStash
+  module PluginMixins
+    module Scheduler
+
+      # @param cron [String] cron-line
+      # @param opts [Hash] scheduler options
+      # @return scheduler instance
+      def start_cron_scheduler!(cron, opts = {}, &block)
+        unless block_given?
+          raise ArgumentError, 'missing block - worker task to execute'
+        end
+        scheduler = new_scheduler(opts)
+        scheduler.schedule_cron(cron, &block)
+        scheduler
+      end
+
+      %w(every at in interval).each do |type|
+        class_eval <<-EVAL, __FILE__, __LINE__ + 1
+          def start_#{type}_scheduler!(arg, opts = {}, &block)
+            unless block_given?
+              raise ArgumentError, 'missing block - worker task to execute'
+            end
+            scheduler = new_scheduler(opts)
+            scheduler.schedule_#{type}(arg, &block)
+            scheduler
+          end
+        EVAL
+      end
+
+      # @param opts [Hash] scheduler options
+      # @return scheduler instance
+      def new_scheduler(opts)
+        unless opts.key?(:thread_name)
+          unless self.class.name
+            raise ArgumentError, "can not generate a thread_name for anonymous class: #{inspect}"
+          end
+          plugin_name = self.class.name.split('::').last # e.g. "jdbc"
+          opts[:thread_name] = "[#{id}]|#{self.class.plugin_type}|#{plugin_name}|scheduler"
```

Comment:
use `pipeline_id = (respond_to?(:execution_context) && execution_context&.pipeline_id) || 'main'`
instead of the plugin's `id` here?

Reply:
I would opt for `pipeline_id`, using the plugin's `id` is not of much help in debugging, when commonly the pipelines definitions doesn't customize plugin's `id`s